docs: security notice (#10555)
langflow-ai/langflow: docs/docs/Support/security.mdx

---
title: Security notice
slug: /security
---

The Langflow UI is an IDE and code execution platform, which means that Langflow is inherently capable of executing arbitrary, developer-provided code.
By design, the Langflow UI includes a code editor allowing developers to author and execute arbitrary Python with full access to the host Langflow backend process, filesystem, and network.
In addition, various components wrap and make use of code execution, in some cases using AI models to generate the code to be executed based on user input.

Given these product features, the Langflow application makes no attempt to prevent malicious activity between different users of a given Langflow process, or to secure the execution environment of the Langflow process, such as preventing access to portions of the local disk. To the extent that limitations are imposed on the access provided to different users, such as limiting flow visibility to a user's flows, these limitations are intended to improve usability, not to impose secure access policies.
For example, users of Langflow can always access the underlying database connection directly.

> The security of code and flows written using Langflow is the developer’s responsibility.

Reviewer: "The developer" is not clear whether it means LF developers or the reader. I suggest being more specific:

```suggestion
The security of code and flows written using Langflow is your responsibility as a developer.
```

> Langflow does not and can not prevent developers from building insecure applications.

Reviewer: I agree with what coderabbit said unless this text is from legal and has to be like this.

> In the same way that Python can’t prevent developers from writing services with SQL-injection vulnerabilities, Langflow can’t prevent developers from writing flows that leak user’s PII, fail to sanitize input, etc.

Reviewer: ```suggestion
In the same way that Python can’t prevent developers from writing services with SQL-injection vulnerabilities, Langflow can’t prevent developers from writing flows that leak PII, fail to sanitize input, or include other insecure logic.
```

> Langflow makes efforts to ensure that components and flows are secure by default and behave as expected, but application behavior is ultimately the developer’s responsibility. 

Reviewer: This statement sounds like a developer is responsible for how the Langflow application acts. It makes sense in a way, but it's not really conveyed clearly. It's also not clear if "THE developer" means LF devs or the reader.
Maybe something like this?

```suggestion
Langflow makes efforts to ensure that components and flows are secure by default and behave as expected, but the security of Langflow in your application logic is ultimately your responsibility as a developer. 
```

Langflow’s cloud security flows from infrastructure.
Cross-tenant isolation is provided by isolating instances of Langflow to individual tenants, not by the Langflow application itself.
Within a Langflow cloud container, users are given full access to the process, disk, and network.

> Security restrictions are enforced between cloud containers, and between containers and the host instance and network.

Reviewer: ```
Security restrictions are enforced between cloud containers, and between containers and the host instance and network.
```

Does _LF_ enforce this or is this enforced by container orchestration/cluster management?

Authentication and authorization are provided and enforced outside of the Langflow container.

> To the extent that services are shared across tenants, such as a shared PostgreSQL database, access limitations are enforced externally, for example by providing DB credentials associated with external security policies.

Reviewer: ```suggestion
To the extent that services are shared across tenants, such as a shared PostgreSQL database, access limitations are enforced externally, for example by providing database credentials associated with external security policies.
```
